Dispatch App

andrewkaufman opened the pull request:
I've added a new app called `dispatch` which can be used from the commandline or via a new `DispatchDialogue` to dispatch graphs. It can take an existing script and a list of nodes to dispatch, or it can be given node classes and construct the script on-the-fly.

ecbcba6 is a bit hacky, but I've discussed it with @johnhaddon and we think its the best way to go for now. I'd also like to avoid 83ba507 if we can think of a way to wrap that up outside the gui app startup.

andrewkaufman commented on python/GafferDispatchUI/DispatchDialogue.py:
```diff
@@ -0,0 +1,261 @@
+
+		self.__script = script
+		# hold a reference to the script window so plugs which launch child windows work properly
+		self.__scriptWindow = GafferUI.ScriptWindow.acquire( script )
```
I also don't like needing this one, but it seemed the easiest way to ensure color swatches and ramps worked as expected.

andrewkaufman replied:
Addressed in 57a006132da5bfa4a0ccfce3d710a6a17249a467

johnhaddon commented on startup/dispatch/project.py:
```diff
@@ -0,0 +1,80 @@
+#
+##########################################################################
+
+## \todo : This entire file was copied directly from startup/gui/project.py
```
Having the same settings seems like a good idea, at least until we get some experience with the new app in practice. What I did for `startup/view/ocio.py` was to just symlink it to `startup/gui/ocio.py`. I'm not sure that's great, but I think it's preferable to a copy that'll inevitably get out of sync.

andrewkaufman replied:
I should clarify, it’s not an exact copy. All the code here does exist in the gui file, but that one has other settings we don’t need here (bookmarks, image format) and some that we explicitly want the opposite of (local executeInBackground userDefault).

What I would like to be consistent across all apps is the registered dispatchers and the location of their job directories.

To do that, I’d need the project variables on the script to be consistent (and exist) as well, or to move the default job directories somewhere else inside ~/gaffer

andrewkaufman replied:
I guess I made this comment after you signed off, or it got buried in the slew of other comments. What are your thoughts now you know its not an exact copy of the gui file?

johnhaddon replied:
I'm still inclined to just symlink. It seems like the bookmarks and image format should be harmless and might even be useful (it does look like the bookmarks want a ScriptWindow, but that could be fixed). The "executeInBackground" default doesn't seem project-related at all really, so we could move it somewhere else in the gui startup, maybe "startup/gui/localDispatcher.py"?

andrewkaufman replied:
Still todo. Though that symlink suggestion seems quite fragile now I think about it... any new setting accicentally put in "startup/gui/proect.py" would affect the dispatch app, possibly without the committer realizing it.

andrewkaufman commented on apps/dispatch/dispatch-1.py:
````diff
@@ -0,0 +1,275 @@
+			Example usage :
+
+			```
+			gaffer dispatch -script comp.gfr -nodes ImageWriter -dispatcher Local -settings -dispatcher.executeInBackground 1
````
I’ll swap that for frameRange or something. I just wanted to show that you can set dispatcher plugs.

I’m hesitant to hide it in the gui though. Is there a use case outside the dispatch app where some code calls launches the dialogue then continues on to do other things while the execution is happening?

johnhaddon replied:
> I’ll swap that for frameRange or something.

Perfect - that sounds like a much more useful example.

> I’m hesitant to hide it in the gui though. Is there a use case outside the dispatch app where some code calls launches the dialogue then continues on to do other things while the execution is happening?

I can see it being useful in a DispatchDialogue embedded in the main gui app, via Jabuka for instance. But for the standalone `dispatch` app I really can't see a valid use. Not a biggy though...

johnhaddon replied:
Actually, maybe it'd be useful in standalone `dispatch` if we added the option to keep the dialogue open and make more dispatches. Let's not worry about hiding this for now...

andrewkaufman replied:
Addressed in 38ad464a325c8f111aedf2b673febd2986c5c908

andrewkaufman commented on apps/dispatch/dispatch-1.py:
```diff
@@ -0,0 +1,275 @@
+
+		dispatcherType = args["dispatcher"].value or GafferDispatch.Dispatcher.getDefaultDispatcherType()
+
+		if args["gui"].value :
```
I originally wrote it (both this block and the constructor) the way you’re suggesting and then changed it to this when I realized we needed the drop down of dispatchers in the dialogue.

I don’t think it’s reasonable to require the client of DispatchDialogue to construct all the dispatchers upfront. That defeats the purpose of registering dispatchers, and in the majority of use cases we have today, it’s perfectly valid for the end user to choose any dispatcher.

It’s also not very friendly for the client of dispatch app. To make the gui work the command would be something like:

gaffer dispatch -gui -nodes SystemCommand -dispatchers Local Tractor

and we’d use the first dispatcher as the default. But then if you omitted the -gui what would we do? Surely not dispatch with both. Erroring for giving two things to a parameter with a plural name doesn’t seem right either...

One thing I realized I’m missing currently is applying the command line dispatcher settings to all dispatchers... currently the frameRange, for instance, can be out of sync, which probably isn’t desirable. I realize this would manifest itself as yet another dialogue constructor arg, but I think it’s probably the right thing to do.

johnhaddon replied:
> I don’t think it’s reasonable to require the client of DispatchDialogue to construct all the dispatchers upfront.

It's totally reasonable; it's about giving control to the client code. For convenience it could have a default value which means "use all the registered dispatchers".

> in the majority of use cases we have today, it’s perfectly valid for the end user to choose any dispatcher.

Do you have a plan for the minority of cases?

> It’s also not very friendly for the client of dispatch app. To make the gui work the command would be something like:

> gaffer dispatch -gui -nodes SystemCommand -dispatchers Local Tractor

Again, the default would be "show 'em all". The command line argument could remain singular (`-dispatcher`), and constrain the UI to show only that dispatcher.

> I realize this would manifest itself as yet another dialogue constructor arg, but I think it’s probably the right thing to do.

The fact that it requires another dialogue constructor arg is a sign that the dialogue constructor isn't right. But I agree, we should apply the frame range to all the dispatchers. We should just do it outside the dialogue...

andrewkaufman replied:
> It's totally reasonable; it's about giving control to the client code.

@lento was actually in favor of your approach originally, but thinking of a future where the dispatchers are in the graph, and the tool author would drop down the ones that should be available, and expose a switch if they wanted to offer a choice. But that future is a ways off, and kind of implies we'd remove the entire Dispatcher tab, only exposing the plugs that were promoted intentionally.

And then I mentioned that if we added a 3rd dispatcher, he'd have to go updating all his tools to account for it, and he started seeing the value in my approach.

But lets say for the moment I agreed to yours...

> For convenience it could have a default value which means "use all the registered dispatchers".

Ok, sounds good.

> The command line argument could remain singular (-dispatcher), and constrain the UI to show only that dispatcher.

What about a command line invocation that wants to give users a choice of dispatchers?

>  But I agree, we should apply the frame range to all the dispatchers. We should just do it outside the dialogue...

So how would a client of the dialogue use both the default "use all the registered dispatchers" arg and also set the frame range on all of them to match? Seems like we still need the extra constructor arg to me.

---------------------

And back to my approach...

> it's about giving control to the client code.

Control comes with a maintenance burden.

> Do you have a plan for the minority of cases?

Trust the end user to be reasonable, rather than tying their hands to one dispatcher or the other.

andrewkaufman replied:
I feel like my last reply comes off more argumentative than I intended it. I’m willing to add the ‘client provides all the dispatchers’ feature if we can come up with an approach that also enables the ‘client provides minimal defaults and still gets all the other dispatchers’ workflow. That 2nd approach would have to work for both clients of the dialogue and clients of the commandline app to be useful at IE though.

johnhaddon replied:
> What about a command line invocation that wants to give users a choice of dispatchers, but also specify a default dispatcher that doesn’t match the registered default dispatcher?

I guess that would require extra arguments. Is this something we need right now?

> So how would a client of the dialogue use both the default "use all the registered dispatchers" arg and also set the frame range on all of them to match? Seems like we still need the extra constructor arg to me.

Hmm, true. Very good point. But I'm afraid this doesn't push me towards wanting the dialogue to construct the dispatchers internally.

> Control comes with a maintenance burden.

Poor APIs come with a maintenance burden. If the dialogue constructs the dispatchers internally, we're going to have to tweak the API every time we find we need more control over that construction. We have perfectly good APIs for querying available dispatchers, creating them and setting their plugs. The permutations are many; I predict a mess if we try to hide them inside the dialogue and add more and more parameters as we find we need more control.

> Trust the end user to be reasonable, rather than tying their hands to one dispatcher or the other.

Personally I'm a fan of that approach, but we're always being asked to police problems via technology. And that was just a single example.

> I’m willing to add the ‘client provides all the dispatchers’ feature if we can come up with an approach that also enables the ‘client provides minimal defaults and still gets all the other dispatchers’ workflow.

Great! I propose the following :

- The DispatchDialogue constructor takes a list of already-constructed dispatchers as I've proposed. It doesn't have any friendly fallback to creating them all, since as you point out, that won't be very useful. This gives us flexibility and a simpler core API but with less convenience.
- We add a utility method somewhere which provides the convenience you're looking for. Perhaps a static `DispatchDialogue.createWithDefaultDispatchers()` method, or a `DispatcherAlgo::createAll()` method. I'll leave the details to you.

How does that sound?

andrewkaufman replied:
I’ll try that approach (probably tomorrow) and see how it feels.

>> What about a command line invocation that wants to give users a choice of dispatchers, but also specify a default dispatcher that doesn’t match the registered default dispatcher?
> I guess that would require extra arguments. Is this something we need right now?

Yes, we need this now. The default is Local, I’d like to keep it that way at IE, but I’d like quite a few of our command line invocations to start the gui with Farm selected and with Local available if the user sees fit to use it.

So what’s the plan for tackling this use case? It shouldn’t affect your design for DispatchDialogue, but will impact the app parameters in some way.

johnhaddon replied:
It seems there's going to be a bit of conflict between the gui and non-gui flavours here whatever we do, because multiple dispatchers don't make sense for the non-gui flavour. I like the way you have it already, which is that a singular `-dispatcher` argument specifies the dispatcher used for the non-gui mode, and also the default dispatcher for gui mode. So maybe we just need an `-otherDispatchers` argument that is used to specify additional dispatchers to choose from in the gui mode? We could allow wildcards in that parameter so that `*` could be used to use all available dispatchers, without needing to know their names.

andrewkaufman replied:
Addressed in 57a006132da5bfa4a0ccfce3d710a6a17249a467, ea8aa5141ae3f7e44d51a06312da06d4ccefd8b1, 38ad464a325c8f111aedf2b673febd2986c5c908, 78549754d2099470c22159585b44606805cd7656

andrewkaufman commented on apps/dispatch/dispatch-1.py:
```diff
@@ -0,0 +1,275 @@
+			return 1
+
+		try :
+			plug.setValue( eval( value ) )
```
I agree it’s fiddly, and it definitely tripped me up when writing the tests. I took a brief look into t, and seemed to be quickly delving into the guts of IECore.ParameterParser, and bailed out...

Unless you have a solid idea how to tackle it, I’m inclned to add the todo and fixup the help example with double quotes.

johnhaddon replied:
> Unless you have a solid idea how to tackle it, I’m inclned to add the todo and fixup the help example with double quotes.

That's fine with me. I don't think a better parser is particularly tricky (it'd be a lot like ParameterParser, but for Plugs), but it is a distraction from the main task. I do want to move Applications to use Plugs for command line arguments, so maybe that's the right time to handle the parsing...

andrewkaufman replied:
Addressed in 38ad464a325c8f111aedf2b673febd2986c5c908

andrewkaufman commented on python/GafferDispatchUI/DispatchDialogue.py:
```diff
@@ -0,0 +1,261 @@
+## A dialogue which can be used to dispatch tasks
+class DispatchDialogue( GafferUI.Dialogue ) :
+
+	def __init__( self, script, tasks, dispatcherType, applyUserDefaults=False, title="Dispatch Tasks", sizeMode=GafferUI.Window.SizeMode.Manual, **kw ) :
```
Happy to drop the script arg.

From the sounds of your intro to the review, you’re leaning towards always applying userDefaults to dispatchers (which would make me very happy) and in which case I can drop this arg and we can separate the discussion of type vs constructed dispatchers from this issue.

andrewkaufman replied:
Addressed in 57a006132da5bfa4a0ccfce3d710a6a17249a467

andrewkaufman commented on python/GafferDispatchUI/DispatchDialogue.py:
```diff
@@ -0,0 +1,261 @@
+				with GafferUI.ListContainer( spacing=2, borderWidth=2 ) as dispatcherTab :
+					with GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Horizontal, spacing=2, borderWidth=2 ) :
+						GafferUI.Label( "Dispatcher" )
+						self.__dispatchersMenu = GafferUI.MultiSelectionMenu( allowMultipleSelection = False, allowEmptySelection = False )
```
I admittadly haven’t had any coffee yet, but I’m struggling to think of a scenario in the IE pipeline where we’d want to restrict the available dispatchers on a per-dispatch basis. Even something like a long sim can be justifiable to run locally, and something seemingly fast, like a model publish can be justifiable to run on a farm. From our pipeline perspective, I’d be happy to set the default type and trust users to stick to the sensible choice for each scenario

What I have been wanting for a while is a global way to deregister a dispatcher (so IE can remove GafferTractor.TractorDispatcher from the menu and the task nodes).

If you wanted, I could hide the menu when there’s only one registered dispatcher... you might find it a bit odd for the UI experience to change when you install tractor though.

johnhaddon replied:
> What I have been wanting for a while is a global way to deregister a dispatcher (so IE can remove GafferTractor.TractorDispatcher from the menu and the task nodes).

Wait, what? You get Tractor in the menu? I had to hack around for ages the other day until I could test the TractorDispatcher. I thought it had been hidden deliberately...

> I’m struggling to think of a scenario in the IE pipeline where we’d want to restrict the available dispatchers on a per-dispatch basis.

Do we count out licenses properly for local dispatch?

> If you wanted, I could hide the menu when there’s only one registered dispatcher... you might find it a bit odd for the UI experience to change when you install tractor though.

There are definitely arguments both ways, but it's also an odd UI experience to have a huge button that opens a menu with no choices in it. Plus, I still think it's reasonable to want to open the DispatchDialogue with one dispatcher specifically.

andrewkaufman replied:
> Wait, what? You get Tractor in the menu? I had to hack around for ages the other day until I could test the TractorDispatcher. I thought it had been hidden deliberately...

There is a serious of events you have to do to get it (its buried in some IEDo method), which R&D almost never does, but artists do all the time. This is how all those taskNode.dispatcher.tractor plugs leaked into production scenes without us noticing.

andrewkaufman replied:
Addressed in 57a006132da5bfa4a0ccfce3d710a6a17249a467, 3e24ca14ab91f57d5a8466caecffe094c6fad437

andrewkaufman commented on python/GafferDispatchUI/DispatchDialogue.py:
```diff
@@ -0,0 +1,261 @@
+				problems.append( "%d %s%s" % ( count, IECore.Msg.levelAsString( level ).capitalize(), "s" if count > 1 else "" ) )
+
+		if not problems :
+			self.close()
```
I intentionally removed all that logic when porting the message widget over... it seems to be a spider web of options... what if we just don’t close so you can see the success icon, and revisit this if users ask for it?

Though I guess your run it again approach doesn’t work unless we add a 2nd button...

andrewkaufman replied:
@lento tells me we rely on the run-again-with-new-settings workflow here currently, but its mainly a legacy thing, and ideally we'd redesign those redundant tools to be "batchable" at the same time we port them to nodes.

What do you think? Should I keep the dialogue open regardless and have the button take you back to the settings page? Or do I need a series of metadata/options for the various open/close/redo configurations?

johnhaddon replied:
Personally I think I'd like to keep the dialogue open to confirm the successful completion, and possibly to have the option of going again. But that might lead immediately to requests for options based on how we'll want to embed it in Jabuka (that's where all those OpDialogue options came from). I don't mind deferring those decisions until we've rolled out a first version and got feedback from more folks...

andrewkaufman replied:
Still need to do this one

johnhaddon commented on python/GafferDispatchUI/DispatchDialogue.py:
```diff
@@ -61,15 +61,15 @@ def __init__( self, script, tasks, dispatcherType, applyUserDefaults=False, titl
+				with GafferUI.ListContainer( borderWidth=3 ) as dispatcherTab :
 					with GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Horizontal, spacing=2, borderWidth=2 ) :
-						GafferUI.Label( "Dispatcher" )
+						GafferUI.Label( "<h4>Current Dispatcher</h4>" )
```
"Current dispatcher" is doubly redundant though. I would go with either just "Dispatcher" or no label.

andrewkaufman replied:
Addressed in 57a006132da5bfa4a0ccfce3d710a6a17249a467

johnhaddon commented on python/GafferDispatchUI/DispatchDialogue.py:
```diff
@@ -61,15 +61,15 @@ def __init__( self, script, tasks, dispatcherType, applyUserDefaults=False, titl
 						self.__dispatchersMenuChanged = self.__dispatchersMenu.selectionChangedSignal().connect( Gaffer.WeakMethod( self.__dispatcherChanged ) )
 
-					self.__dispatcherFrame = GafferUI.Frame( borderStyle=GafferUI.Frame.BorderStyle.None, borderWidth=0 )
+					self.__dispatcherFrame = GafferUI.Frame( borderWidth=2 )
```
I think the goal should be to make the dispatcher's NodeUI line up with the other NodeUIs in general, and to have the appearance of the same number of nesting levels on each tab.

andrewkaufman replied:
Is that even possible unless we totally ditch the dropdown?

johnhaddon replied:
It should be I think. Containers with zero border shouldn't indent their contents at all, so it should be possible to have a dropdown and still have the NodeUIs line up in terms of left/right/bottom. The top will obviously be pushed down though...

andrewkaufman replied:
Addressed in 57a006132da5bfa4a0ccfce3d710a6a17249a467

andrewkaufman commented on apps/dispatch/dispatch-1.py:
````diff
@@ -0,0 +1,275 @@
+			```
+			gaffer dispatch -script comp.gfr -nodes ImageWriter -dispatcher Local -settings -dispatcher.executeInBackground 1
+
+			gaffer dispatch -gui -nodes GafferDispatch.SystemCommand -dispatcher Local -settings -SystemCommand.command "ls -l"
````
Addressed in 38ad464a325c8f111aedf2b673febd2986c5c908

andrewkaufman commented on apps/dispatch/dispatch-1.py:
```diff
@@ -0,0 +1,275 @@
+	@staticmethod
+	def __create( path ) :
+
+		path = path.split( "." )
```
Addressed in 8c1312abadc2b0dd453ce4921754e03a8dbcd524. Note I didn't add that `GafferDispatch` convenience, because I don't want to make it easy to forget that its inconvenient for all IE nodes.

johnhaddon replied:
This is still giving nasty errors for some values :

```
> gaffer dispatch -nodes GafferDispatch

gaffer Traceback (most recent call last):
  File "/disk1/john/dev/build/gaffer/bin/gaffer.py", line 169, in <module>
    result = app.run()
  File "/disk1/john/dev/build/gaffer/python/Gaffer/Application.py", line 110, in run
    return self.__run()
  File "/disk1/john/dev/build/gaffer/python/Gaffer/Application.py", line 142, in __run
    return self._run( self.parameters().getValidatedValue() )
  File "/disk1/john/dev/build/gaffer/apps/dispatch/dispatch-1.py", line 164, in _run
    node = self.__create( nodeName )
  File "/disk1/john/dev/build/gaffer/apps/dispatch/dispatch-1.py", line 294, in __create
    return result()
TypeError: 'module' object is not callable
```

```
> gaffer dispatch -nodes GafferDispatch.Oops

Traceback (most recent call last):
  File "/disk1/john/dev/build/gaffer/bin/gaffer.py", line 169, in <module>
    result = app.run()
  File "/disk1/john/dev/build/gaffer/python/Gaffer/Application.py", line 110, in run
    return self.__run()
  File "/disk1/john/dev/build/gaffer/python/Gaffer/Application.py", line 142, in __run
    return self._run( self.parameters().getValidatedValue() )
  File "/disk1/john/dev/build/gaffer/apps/dispatch/dispatch-1.py", line 164, in _run
    node = self.__create( nodeName )
  File "/disk1/john/dev/build/gaffer/apps/dispatch/dispatch-1.py", line 292, in __create
    result = getattr( result, n )
AttributeError: 'module' object has no attribute 'Oops'
```

andrewkaufman commented on python/GafferDispatchUI/DispatchDialogue.py:
```diff
@@ -0,0 +1,261 @@
+		self.__tabs.setCurrent( self.__tabs[0] )
+		self._getWidget().setChild( self.__settings )
+
+	def setTasks( self, tasks ) :
```
Addressed in 57a006132da5bfa4a0ccfce3d710a6a17249a467

andrewkaufman commented on python/GafferDispatchUI/DispatchDialogue.py:
```diff
@@ -0,0 +1,261 @@
+				Gaffer.Metadata.registerValue( task, "layout:customWidget:dispatchButton:widgetType", "", persistent = False )
+			self.__tabs.insert( 0, nodeFrame, label = task.relativeName( self.__script ) )
+
+	def setDispatcher( self, dispatcherType ) :
```
Addressed in 57a006132da5bfa4a0ccfce3d710a6a17249a467

andrewkaufman commented on python/GafferDispatchUI/DispatchDialogue.py:
```diff
@@ -0,0 +1,261 @@
+
+		self.__script = script
+		# hold a reference to the script window so plugs which launch child windows work properly
+		self.__scriptWindow = GafferUI.ScriptWindow.acquire( script )
```
Addressed in 57a006132da5bfa4a0ccfce3d710a6a17249a467

andrewkaufman commented:
Ok, I think I've got everything except keeping the gui open on success and deciding if we're going to generalize the project settings across both apps. I'll try to get those done Monday.

andrewkaufman commented:
6b3251f adds the ability to keep the dialogue open or close it.

andrewkaufman commented:
> "INFO : RESULT" message

Addressed in 7cd76ea332515a0f67ed5a5cbfc0a30af4267d3f

> symlink suggestion

Addressed in 192e23154ad07bb9cf1b3e07b277b37a6b6eb8fc, c81d4b032ca7edae8d4e2ded1eaba38abbfe3345, and 13584688af45249a485a490e9eec96697d6cb39b

> nasty errors

Addressed in 7f8f858b3d8eefc1540ee293826310ec3632439b

> local foreground dispatch which fails

Not yet addressed, but I agree, this does seem like a problem... How do you recommend fixing it in LocalDispatcher?

johnhaddon commented:
> > local foreground dispatch which fails
> Not yet addressed, but I agree, this does seem like a problem... How do you recommend fixing it in LocalDispatcher?

I think we want to allow the exception thrown by `batch.execute()` to percolate back up out of `LocalDispatcher.dispatch()`, instead of suppressing it. I'm unsure of the details of what to do with the `__reportFailed()` stuff we're doing at the same time as suppressing, because I don't know who's listening (is the Job/JobPool stuff actually relevant to a foreground dispatch?). Either we rethrow the exception I guess, or we just don't bother catching it in the first place...

andrewkaufman commented:
> local foreground dispatch which fails

Addressed in f793784 and 167fcaa. The latter improves the error dialogue that's displayed in the gui app for the same broken SystemCommand.

> I'm unsure of the details of what to do with the __reportFailed() stuff we're doing at the same time as suppressing, because I don't know who's listening (is the Job/JobPool stuff actually relevant to a foreground dispatch?).

Those jobs do display in the _LocalJobsWindow, although prior to these commits the jobs displayed in a semi-broken state. Now they display as failed and include the traceback in the debug messages.

johnhaddon commented on python/GafferDispatchTest/LocalDispatcherTest.py:
```diff
@@ -492,7 +492,11 @@ def testFailure( self ) :
+		try :
+			dispatcher.dispatch( [ s["n1"] ] )
+		except Exception as e :
+			self.assertTrue( "No such file or directory" in str(e) )
```
Nitpick : this is what `assertRaisesRegexp()` is for...

andrewkaufman replied:
Addressed in 5d9a43d2c2bb335682bf02ba14509b6c20baacc2

andrewkaufman commented:
> are we ready to merge from your point of view?

All good from my perspective. I may have to come back to this once @ivanimanishi starts using it in Jabuka, but I'd rather merge as is and address his needs in a future PR.
